### PR TITLE
[FLINK-14337][hs] Better handling of failed fetches

### PR DESCRIPTION
**Pull request by zentol**
Some small adjustments to the HistoryServer to prevent NPEs when processing empty/partial/corrupted archives, and ensure that archive fetches are retried unless the fetch was completed successfully.

Tests weren't added intentionally. The NPE issue is difficult to test without exposing details about the JSON structure to the test, while the other issue is simply difficult to test without larger refactorings.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java [hunk 1]
```diff
@@ -112,12 +112,19 @@ public static Path archiveJob(Path rootPath, JobID jobId, Collection<ArchivedJso
 			JsonNode archive = mapper.readTree(output.toByteArray());
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java]: Can't this also be `null`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java [hunk 2]
```diff
@@ -112,12 +112,19 @@ public static Path archiveJob(Path rootPath, JobID jobId, Collection<ArchivedJso
...
+				String json = failOnNull(archivePart.get(JSON)).asText();
 				archives.add(new ArchivedJson(path, json));
 			}
 			return archives;
 		}
 	}
+
+	private static <X> X failOnNull(X value) throws IOException {
+		if (value == null) {
+			throw new IOException("Job archive did not conform to expected format.");
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java]: Maybe we could add which `file` contained the invalid format.

File: flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java (hunk 1 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java]: fixed

File: flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java]: I've reworked this commit (now uses a simple try-catch block) and include the filename in the error message.

**Pull request merged**